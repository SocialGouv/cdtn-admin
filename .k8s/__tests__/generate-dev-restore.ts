//

import { getEnvManifests } from "@socialgouv/kosko-charts/testing";
import { project } from "@socialgouv/kosko-charts/testing/fake/gitlab-ci.env";

jest.setTimeout(1000 * 60);
test("kosko generate --dev jobs/restore", async () => {
  expect(
    await getEnvManifests("dev", "jobs/restore", {
      ...project("cdtn-admin").dev,
    })
  ).toMatchSnapshot();
});
