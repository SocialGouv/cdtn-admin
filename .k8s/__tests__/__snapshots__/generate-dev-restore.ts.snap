// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kosko generate --dev jobs/restore 1`] = `
"---
apiVersion: batch/v1
kind: Job
metadata:
  name: restore-container-master
  namespace: cdtn-admin-secret
  labels:
    application: master-dev42-cdtn-admin
    owner: cdtn-admin
    team: cdtn-admin
    cert: wildcard
    component: restore-master
  annotations:
    kapp.k14s.io/update-strategy: always-replace
spec:
  backoffLimit: 0
  template:
    metadata:
      annotations:
        kapp.k14s.io/deploy-logs: for-new-or-existing
    spec:
      containers:
        - command:
            - sh
            - '-c'
            - >+


              [ ! -z $SOURCE_CONTAINER ] || (echo \\"No SOURCE_CONTAINER\\"; exit 1)

              [ ! -z $SOURCE_ACCOUNT_NAME ] || (echo \\"No SOURCE_ACCOUNT_NAME\\";
              exit 1)

              [ ! -z $SOURCE_ACCOUNT_KEY ] || (echo \\"No SOURCE_ACCOUNT_KEY\\";
              exit 1)

              [ ! -z $DESTINATION_CONTAINER ] || (echo \\"No
              DESTINATION_CONTAINER\\"; exit 1)

              [ ! -z $DESTINATION_ACCOUNT_NAME ] || (echo \\"No
              DESTINATION_ACCOUNT_NAME\\"; exit 1)

              [ ! -z $DESTINATION_ACCOUNT_KEY ] || (echo \\"No
              DESTINATION_ACCOUNT_KEY\\"; exit 1)



              echo \\"starting restore container $SOURCE_CONTAINER into
              $DESTINATION_CONTAINER\\"


              # copy container

              az storage copy -r   --source-account-name
              \\"$SOURCE_ACCOUNT_NAME\\"   --source-account-key
              \\"$SOURCE_ACCOUNT_KEY\\"   --source-container \\"$SOURCE_CONTAINER\\"  
              --account-name \\"$DESTINATION_ACCOUNT_NAME\\"   --account-key
              \\"$DESTINATION_ACCOUNT_KEY\\"   --destination-container
              \\"$DESTINATION_CONTAINER\\"


              # get original permission

              PUBLIC_ACCESS=\`az storage container show-permission  
              --account-name \\"$SOURCE_ACCOUNT_NAME\\"   --account-key
              \\"$SOURCE_ACCOUNT_KEY\\"   --name $SOURCE_CONTAINER | jq -r
              \\".publicAccess\\"\`


              # set original permission

              az storage container set-permission   --name
              $DESTINATION_CONTAINER   --account-name
              \\"$DESTINATION_ACCOUNT_NAME\\"   --account-key
              \\"$DESTINATION_ACCOUNT_KEY\\"   --public-access $PUBLIC_ACCESS

          env:
            - name: SOURCE_ACCOUNT_NAME
              valueFrom:
                secretKeyRef:
                  key: azurestorageaccountname
                  name: azure-cdtnadmindev-volume
            - name: SOURCE_ACCOUNT_KEY
              valueFrom:
                secretKeyRef:
                  key: azurestorageaccountkey
                  name: azure-cdtnadmindev-volume
            - name: DESTINATION_ACCOUNT_NAME
              valueFrom:
                secretKeyRef:
                  key: azurestorageaccountname
                  name: azure-cdtnadminprod-volume
            - name: DESTINATION_ACCOUNT_KEY
              valueFrom:
                secretKeyRef:
                  key: azurestorageaccountkey
                  name: azure-cdtnadminprod-volume
            - name: SOURCE_CONTAINER
              value: cdtn
            - name: DESTINATION_CONTAINER
              value: cdtn
          envFrom: []
          image: mcr.microsoft.com/azure-cli:2.15.1
          imagePullPolicy: IfNotPresent
          name: restore-db
          resources:
            limits:
              cpu: 300m
              memory: 512Mi
            requests:
              cpu: 50m
              memory: 64Mi
      restartPolicy: Never
  ttlSecondsAfterFinished: 86400
---
apiVersion: v1
kind: ConfigMap
data:
  post-restore.sql: |+
    TRUNCATE TABLE \\"auth\\".\\"users\\" CASCADE;

    --
    -- DISABLE TRIGGERS
    --

    ALTER TABLE auth.users DISABLE TRIGGER USER;

    WITH admin_row AS (
    INSERT INTO auth.users (email, PASSWORD, name, default_role, active)
        VALUES ('codedutravailnumerique@travail.gouv.fr', '$argon2i$v=19$m=4096,t=3,p=1$n9eoWSv+5sCgc7SjB5hLig$iBQ7NzrHHLkJSku/dCetNs+n/JI1CMdkWaoZsUekLU8', 'Administrateur', 'admin', TRUE)
      RETURNING
        id, default_role)
      INSERT INTO auth.user_roles (ROLE, user_id)
      SELECT
        default_role,
        id
      FROM
        admin_row;

    WITH admin_row AS (
    INSERT INTO auth.users (email, PASSWORD, name, default_role, active)
        VALUES ('utilisateur@travail.gouv.fr', '$argon2i$v=19$m=4096,t=3,p=1$PqKPf9cxunVLLtEcINHhWQ$CwHKhk71fc8LGp6BWbcFPzQ2ftOiHa7vUkp1eAqVHSM', 'Utilisateur', 'user', TRUE)
      RETURNING
        id, default_role)
      INSERT INTO auth.user_roles (ROLE, user_id)
      SELECT
        default_role,
        id
      FROM
        admin_row;

    --
    -- create empty audit logs table
    --

    CREATE TABLE IF NOT EXISTS audit.logged_actions (
        event_id bigserial primary key,

        schema_name text not null,
        table_name text not null,
        relid oid not null,

        session_user_name text,
        hasura_user jsonb,

        action_tstamp_tx TIMESTAMP WITH TIME ZONE NOT NULL,
        action_tstamp_stm TIMESTAMP WITH TIME ZONE NOT NULL,
        action_tstamp_clock TIMESTAMP WITH TIME ZONE NOT NULL,
        transaction_id bigint,

        application_name text,
        client_addr inet,
        client_port integer,

        client_query text,
        action TEXT NOT NULL CHECK (action IN ('I','D','U', 'T')),
        row_data jsonb,
        changed_fields jsonb,
        statement_only boolean not null
    );

    --
    -- ENABLE TRIGGERS
    --

    ALTER TABLE auth.users ENABLE TRIGGER USER;

    --
    -- Kill all connections !
    -- Make all connected services restart !
    -- Hasura migration will be re-applyed.
    --

    SELECT pg_terminate_backend(pid)
    FROM pg_stat_activity
    WHERE pid <> pg_backend_pid()
    AND datname = 'autodevops_master';


metadata:
  name: post-restore-script-configmap-0123456
  namespace: cdtn-admin-secret
  labels:
    application: master-dev42-cdtn-admin
    owner: cdtn-admin
    team: cdtn-admin
    cert: wildcard
    component: restore-master
---
apiVersion: batch/v1
kind: Job
metadata:
  name: restore-db-master
  namespace: cdtn-admin-secret
  labels:
    application: master-dev42-cdtn-admin
    owner: cdtn-admin
    team: cdtn-admin
    cert: wildcard
    component: restore-master
  annotations:
    kapp.k14s.io/update-strategy: always-replace
spec:
  backoffLimit: 0
  template:
    metadata:
      annotations:
        kapp.k14s.io/deploy-logs: for-new-or-existing
    spec:
      containers:
        - command:
            - sh
            - '-c'
            - >


              echo \\"starting restore into $PGHOST/$PGDATABASE\\"


              [ ! -z $PGDATABASE ] || (echo \\"No PGDATABASE\\"; exit 1)

              [ ! -z $PGHOST ] || (echo \\"No PGHOST\\"; exit 1)

              [ ! -z $PGUSER ] || (echo \\"No PGUSER\\"; exit 1)

              [ ! -z $PGPASSWORD ] || (echo \\"No PGPASSWORD\\"; exit 1)

              [ ! -z $OWNER ] || (echo \\"No OWNER\\"; exit 1)


              # get latest backup folder

              LATEST=$(ls -1Fr /mnt/data | head -n 1);

              DUMP=\\"/mnt/data/\${LATEST}\${FILE}\\"

              echo \\"Restore \${DUMP} into \${PGDATABASE}\\";


              pg_isready;


              pg_restore   --dbname \${PGDATABASE}   --clean --if-exists  
              --exclude-schema=audit   --no-owner   --role \${OWNER}   --no-acl  
              --verbose   \${DUMP};


              psql -v ON_ERROR_STOP=1 \${PGDATABASE} -c \\"ALTER SCHEMA public
              owner to \${OWNER};\\"


              [ -f \\"/mnt/scripts/post-restore.sql\\" ] && psql -v ON_ERROR_STOP=1
              -a < /mnt/scripts/post-restore.sql
          env:
            - name: PGDATABASE
              value: autodevops_master
            - name: OWNER
              value: user_master
            - name: FILE
              value: hasura_prod_db.psql.gz
          envFrom:
            - secretRef:
                name: azure-pg-admin-user-dev
          image: ghcr.io/socialgouv/docker/azure-db:6.0.1
          imagePullPolicy: IfNotPresent
          name: restore-db
          resources:
            limits:
              cpu: 300m
              memory: 512Mi
            requests:
              cpu: 100m
              memory: 64Mi
          volumeMounts:
            - mountPath: /mnt/data
              name: backups
            - mountPath: /mnt/scripts
              name: scripts
      restartPolicy: OnFailure
      volumes:
        - azureFile:
            readOnly: true
            secretName: azure-cdtnadminprod-volume
            shareName: cdtn-admin-backup-restore
          name: backups
        - configMap:
            name: post-restore-script-configmap-0123456
          name: scripts
      initContainers:
        - env:
            - name: PGHOST
              value: cdtnadmindevserver.postgres.database.azure.com
            - name: PGDATABASE
              value: autodevops_master
            - name: PGPASSWORD
              value: password_master
            - name: PGUSER
              value: user_master@cdtnadmindevserver.postgres.database.azure.com
            - name: PGSSLMODE
              value: require
          envFrom: []
          image: ghcr.io/socialgouv/docker/wait-for-postgres:6.0.1
          imagePullPolicy: Always
          name: wait-for-postgres
          resources:
            limits:
              cpu: 20m
              memory: 32Mi
            requests:
              cpu: 5m
              memory: 16Mi
  ttlSecondsAfterFinished: 86400
"
`;
